File: server/src/main/java/com/larsreimann/api_editor/codegen/FunctionAdapterContentBuilder.java
package com.larsreimann.api_editor.codegen;

import com.larsreimann.api_editor.io.FileBuilder;
import com.larsreimann.api_editor.model.AnnotatedPythonFunction;
import com.larsreimann.api_editor.model.AnnotatedPythonParameter;
import com.larsreimann.api_editor.model.Boundary;
import com.larsreimann.api_editor.model.ComparisonOperator;
import com.larsreimann.api_editor.model.PythonParameterAssignment;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class FunctionAdapterContentBuilder extends FileBuilder {
    AnnotatedPythonFunction pythonFunction;

    /**
     * Constructor for FunctionAdapterContentBuilder
     *
     * @param pythonFunction The function whose adapter content should be built
     */
    public FunctionAdapterContentBuilder(
        AnnotatedPythonFunction pythonFunction
    ) {
        this.pythonFunction = pythonFunction;
    }

    /**
     * Builds a string containing the formatted function content
     *
     * @return The string containing the formatted function content
     */
    public String buildFunction() {
        String constructorSuffix = "";
        if (pythonFunction.isConstructor()) {
            String constructorSeparator = "";
            String assignments = listToString(buildAttributeAssignments(), 1);
            if (!assignments.isBlank()) {
                constructorSeparator = "\n";
            }
            constructorSuffix = constructorSeparator + assignments;
        }

        return "def "
            + pythonFunction.getName()
            + "("
            + buildParameters()
            + ")"
            + ":\n"
            + indent(buildFunctionBody() + constructorSuffix);
    }

    private List<String> buildAttributeAssignments() {
        List<String> attributeAssignments = new ArrayList<>();
        for (AnnotatedPythonParameter parameterAttribute : pythonFunction.getParameters()) {
            if (parameterAttribute.getAssignedBy().equals(PythonParameterAssignment.ATTRIBUTE)) {
                attributeAssignments.add(
                    "self."
                        + parameterAttribute.getName()
                        + " = "
                        + parameterAttribute.getDefaultValue()
                );
            }
        }
        return attributeAssignments;
    }

    private String buildParameters() {
        String formattedFunctionParameters = "";
        List<String> implicitParameters = new ArrayList<>();
        List<String> positionOnlyParameters = new ArrayList<>();
        List<String> positionOrNameParameters = new ArrayList<>();
        List<String> nameOnlyParameters = new ArrayList<>();
        pythonFunction.getParameters().forEach(pythonParameter -> {
            switch (pythonParameter.getAssignedBy()) {
                case IMPLICIT -> implicitParameters
                    .add(buildFormattedParameter(pythonParameter));
                case POSITION_ONLY -> positionOnlyParameters
                    .add(buildFormattedParameter(pythonParameter));
                case POSITION_OR_NAME -> positionOrNameParameters
                    .add(buildFormattedParameter(pythonParameter));
                case NAME_ONLY -> nameOnlyParameters
                    .add(buildFormattedParameter(pythonParameter));
            }
        });
        // A function has at most one implicit parameter
        assert implicitParameters.size() < 2;
        boolean hasImplicitParameter = !implicitParameters.isEmpty();
        boolean hasPositionOnlyParameters = !positionOnlyParameters.isEmpty();
        boolean hasPositionOrNameParameters = !positionOrNameParameters.isEmpty();
        boolean hasNameOnlyParameters = !nameOnlyParameters.isEmpty();

        if (hasImplicitParameter) {
            formattedFunctionParameters =
                formattedFunctionParameters
                    + implicitParameters.get(0);
            if (
                hasPositionOnlyParameters ||
                    hasPositionOrNameParameters ||
                    hasNameOnlyParameters
            ) {
                formattedFunctionParameters =
                    formattedFunctionParameters + ", ";
            }
        }

        if (hasPositionOnlyParameters) {
            formattedFunctionParameters =
                formattedFunctionParameters
                    + String.join(", ", positionOnlyParameters);
            if (hasPositionOrNameParameters) {
                formattedFunctionParameters =
                    formattedFunctionParameters
                        + ", /, ";
            } else if (hasNameOnlyParameters) {
                formattedFunctionParameters =
                    formattedFunctionParameters
                        + ", /";
            } else {
                formattedFunctionParameters =
                    formattedFunctionParameters
                        + ", /";
            }
        }
        if (hasPositionOrNameParameters) {
            formattedFunctionParameters =
                formattedFunctionParameters
                    + String.join(", ", positionOrNameParameters);
        }
        if (hasNameOnlyParameters) {
            if (hasPositionOnlyParameters || hasPositionOrNameParameters) {
                formattedFunctionParameters =
                    formattedFunctionParameters + ", *, ";
            } else {
                formattedFunctionParameters =
                    formattedFunctionParameters + "*, ";
            }
            formattedFunctionParameters =
                formattedFunctionParameters
                    + String.join(", ", nameOnlyParameters);
        }
        return formattedFunctionParameters;
    }

    private String buildFormattedParameter(AnnotatedPythonParameter pythonParameter) {
        String formattedParameter = pythonParameter.getName();
        String defaultValue = pythonParameter.getDefaultValue();
        if (defaultValue != null) {
            formattedParameter = formattedParameter + "=" + defaultValue;
        }
        return formattedParameter;
    }

    private String buildFunctionBody() {
        String formattedBoundaries = listToString(buildBoundaryChecks(), 1);
        if (!formattedBoundaries.isBlank()) {
            formattedBoundaries = formattedBoundaries + "\n";
        }
        return formattedBoundaries
            + Objects.requireNonNull(pythonFunction.getOriginalDeclaration()).getQualifiedName()
            + "("
            + buildParameterCall()
            + ")";
    }

    private List<String> buildBoundaryChecks() {
        List<String> formattedBoundaries = new ArrayList<>();
        for (AnnotatedPythonParameter pythonParameter :
            pythonFunction
                .getParameters()
                .stream()
                .filter(pythonParameter -> pythonParameter.getBoundary() != null).toList()) {
            Boundary boundary = pythonParameter.getBoundary();
            assert boundary != null;
            if (boundary.isDiscrete()) {
                formattedBoundaries.add(
                    "if not (isinstance("
                        + pythonParameter.getName()
                        + ", int) or (isinstance("
                        + pythonParameter.getName()
                        + ", float)"
                        + " and "
                        + pythonParameter.getName()
                        + ".is_integer())):\n"
                        + indent(
                        "raise ValueError('"
                            + pythonParameter.getName()
                            + " needs to be an integer, but {} was assigned."
                            + "'.format("
                            + pythonParameter.getName()
                            + "))"
                    )
                );
            }

            if (boundary.getLowerLimitType() != ComparisonOperator.UNRESTRICTED && boundary.getUpperLimitType() != ComparisonOperator.UNRESTRICTED) {
                formattedBoundaries.add(
                    "if not "
                        + boundary.getLowerIntervalLimit()
                        + " "
                        + boundary.getLowerLimitType().getOperator()
                        + " "
                        + pythonParameter.getName()
                        + " "
                        + boundary.getUpperLimitType().getOperator()
                        + " "
                        + boundary.getUpperIntervalLimit()
                        + ":\n"
                        + indent(
                            "raise ValueError('Valid values of "
                            + pythonParameter.getName()
                            + " must be in "
                            + boundary.asInterval()
                            + ", but {} was assigned."
                            + "'.format("
                            + pythonParameter.getName()
                            + "))"
                    )
                );
            } else if (boundary.getLowerLimitType() == ComparisonOperator.LESS_THAN) {
                formattedBoundaries.add(
                    "if not "
                        + boundary.getLowerIntervalLimit()
                        + " < "
                        + pythonParameter.getName()
                        + ":\n"
                        + indent(
                        "raise ValueError('Valid values of "
                            + pythonParameter.getName()
                            + " must be greater than "
                            + boundary.getLowerIntervalLimit()
                            + ", but {} was assigned."
                            + "'.format("
                            + pythonParameter.getName()
                            + "))"
                    )
                );
            } else if (boundary.getLowerLimitType() == ComparisonOperator.LESS_THAN_OR_EQUALS) {
                formattedBoundaries.add(
                    "if not "
                        + boundary.getLowerIntervalLimit()
                        + " <= "
                        + pythonParameter.getName()
                        + ":\n"
                        + indent(
                        "raise ValueError('Valid values of "
                            + pythonParameter.getName()
                            + " must be greater than or equal to "
                            + boundary.getLowerIntervalLimit()
                            + ", but {} was assigned."
                            + "'.format("
                            + pythonParameter.getName()
                            + "))"
                    )
                );
            }
            if (boundary.getUpperLimitType() == ComparisonOperator.LESS_THAN) {
                formattedBoundaries.add(
                    "if not "
                        + pythonParameter.getName()
                        + " < "
                        + boundary.getUpperIntervalLimit()
                        + ":\n"
                        + indent(
                        "raise ValueError('Valid values of "
                            + pythonParameter.getName()
                            + " must be less than "
                            + boundary.getUpperIntervalLimit()
                            + ", but {} was assigned."
                            + "'.format("
                            + pythonParameter.getName()
                            + "))"
                    )
                );
            } else if (boundary.getUpperLimitType() == ComparisonOperator.LESS_THAN) {
                formattedBoundaries.add(
                    "if not "
                        + pythonParameter.getName()
                        + " <= "
                        + boundary.getUpperIntervalLimit()
                        + ":\n"
                        + indent(
                        "raise ValueError('Valid values of "
                            + pythonParameter.getName()
                            + " must be less than or equal to "
                            + boundary.getUpperIntervalLimit()
                            + ", but {} was assigned."
                            + "'.format("
                            + pythonParameter.getName()
                            + "))"
                    )
                );
            }
        }
        return formattedBoundaries;
    }

    private String buildParameterCall() {
        List<String> formattedParameters = new ArrayList<>();
        Map<String, String> originalNameToValueMap = new HashMap<>();
        pythonFunction.getParameters().forEach(
            pythonParameter -> {
                String value;
                if (
                    pythonParameter.getAssignedBy().equals(PythonParameterAssignment.CONSTANT)
                        || pythonParameter.getAssignedBy().equals(PythonParameterAssignment.ATTRIBUTE)
                ) {
                    value = pythonParameter.getDefaultValue();
                } else {
                    value = pythonParameter.getName();
                }
                originalNameToValueMap.put(
                    Objects
                        .requireNonNull(pythonParameter.getOriginalDeclaration())
                        .getName(),
                    value
                );
            }
        );
        Objects.requireNonNull(pythonFunction.getOriginalDeclaration())
            .getParameters().stream()
            .filter(pythonParameter ->
                !pythonParameter.getAssignedBy()
                    .equals(PythonParameterAssignment.IMPLICIT)
            ).forEach(pythonParameter -> {
                if (pythonParameter.getAssignedBy()
                    == PythonParameterAssignment.NAME_ONLY) {
                    formattedParameters.add(
                        pythonParameter.getName()
                            + "="
                            + originalNameToValueMap.get(pythonParameter.getName())
                    );
                } else {
                    formattedParameters.add(
                        originalNameToValueMap.get(pythonParameter.getName())
                    );
                }
            });
        return String.join(", ", formattedParameters);
    }
}

File: server/src/main/kotlin/com/larsreimann/api_editor/codegen/FunctionAdapterContentBuilder.kt
package com.larsreimann.api_editor.codegen

import com.larsreimann.api_editor.model.ComparisonOperator
import com.larsreimann.api_editor.model.PythonParameterAssignment
import com.larsreimann.api_editor.model.SerializablePythonParameter
import com.larsreimann.api_editor.mutable_model.MutablePythonFunction
import com.larsreimann.api_editor.mutable_model.MutablePythonParameter
import java.util.Objects
import java.util.function.Consumer

/**
 * Constructor for FunctionAdapterContentBuilder
 *
 * @param pythonFunction The function whose adapter content should be built
 */
class FunctionAdapterContentBuilder(var pythonFunction: MutablePythonFunction) {

    /**
     * Builds a string containing the formatted function content
     *
     * @return The string containing the formatted function content
     */
    fun buildFunction(): String {
        var constructorSuffix = ""
        if (pythonFunction.isConstructor()) {
            var constructorSeparator = ""
            val assignments = listToString(buildAttributeAssignments(), 1)
            if (assignments.isNotBlank()) {
                constructorSeparator = "\n"
            }
            constructorSuffix = constructorSeparator + assignments
        }
        return """
              def ${pythonFunction.name}(${buildParameters()}):
              ${indent(buildFunctionBody() + constructorSuffix)}
              """.trimIndent()
    }

    private fun buildAttributeAssignments(): List<String> {
        val attributeAssignments: MutableList<String> = ArrayList()
        for ((name, defaultValue, assignedBy) in pythonFunction.parameters) {
            if (assignedBy == PythonParameterAssignment.ATTRIBUTE) {
                attributeAssignments.add(
                    "self."
                        + name
                        + " = "
                        + defaultValue
                )
            }
        }
        return attributeAssignments
    }

    private fun buildParameters(): String {
        var formattedFunctionParameters = ""
        val implicitParameters: MutableList<String> = ArrayList()
        val positionOnlyParameters: MutableList<String> = ArrayList()
        val positionOrNameParameters: MutableList<String> = ArrayList()
        val nameOnlyParameters: MutableList<String> = ArrayList()
        pythonFunction.parameters.forEach(Consumer { pythonParameter: MutablePythonParameter ->
            when (pythonParameter.assignedBy) {
                PythonParameterAssignment.IMPLICIT -> implicitParameters
                    .add(buildFormattedParameter(pythonParameter))
                PythonParameterAssignment.POSITION_ONLY -> positionOnlyParameters
                    .add(buildFormattedParameter(pythonParameter))
                PythonParameterAssignment.POSITION_OR_NAME -> positionOrNameParameters
                    .add(buildFormattedParameter(pythonParameter))
                PythonParameterAssignment.NAME_ONLY -> nameOnlyParameters
                    .add(buildFormattedParameter(pythonParameter))
                else -> {}
            }
        })
        assert(implicitParameters.size < 2)
        val hasImplicitParameter = implicitParameters.isNotEmpty()
        val hasPositionOnlyParameters = positionOnlyParameters.isNotEmpty()
        val hasPositionOrNameParameters = positionOrNameParameters.isNotEmpty()
        val hasNameOnlyParameters = nameOnlyParameters.isNotEmpty()
        if (hasImplicitParameter) {
            formattedFunctionParameters = (formattedFunctionParameters
                + implicitParameters[0])
            if (hasPositionOnlyParameters ||
                hasPositionOrNameParameters ||
                hasNameOnlyParameters
            ) {
                formattedFunctionParameters = "$formattedFunctionParameters, "
            }
        }
        if (hasPositionOnlyParameters) {
            formattedFunctionParameters = (formattedFunctionParameters
                + java.lang.String.join(", ", positionOnlyParameters))
            formattedFunctionParameters = if (hasPositionOrNameParameters) {
                (formattedFunctionParameters
                    + ", /, ")
            } else if (hasNameOnlyParameters) {
                (formattedFunctionParameters
                    + ", /")
            } else {
                (formattedFunctionParameters
                    + ", /")
            }
        }
        if (hasPositionOrNameParameters) {
            formattedFunctionParameters = (formattedFunctionParameters + java.lang.String.join(", ", positionOrNameParameters))
        }
        if (hasNameOnlyParameters) {
            formattedFunctionParameters = if (hasPositionOnlyParameters || hasPositionOrNameParameters) {
                "$formattedFunctionParameters, *, "
            } else {
                "$formattedFunctionParameters*, "
            }
            formattedFunctionParameters = (formattedFunctionParameters
                + java.lang.String.join(", ", nameOnlyParameters))
        }
        return formattedFunctionParameters
    }

    private fun buildFormattedParameter(pythonParameter: MutablePythonParameter): String {
        var formattedParameter = pythonParameter.name
        val defaultValue = pythonParameter.defaultValue
        if (defaultValue != null) {
            formattedParameter = "$formattedParameter=$defaultValue"
        }
        return formattedParameter
    }

    private fun buildFunctionBody(): String {
        var formattedBoundaries = listToString(buildBoundaryChecks(), 1)
        if (formattedBoundaries.isNotBlank()) {
            formattedBoundaries = """
                $formattedBoundaries

                """.trimIndent()
        }
        return (formattedBoundaries
            + Objects.requireNonNull(pythonFunction.originalDeclaration)!!.qualifiedName
            + "("
            + buildParameterCall()
            + ")")
    }

    private fun buildBoundaryChecks(): List<String> {
        val formattedBoundaries: MutableList<String> = ArrayList()
        pythonFunction
            .parameters
            .filter { (_, _, _, _, _, _, boundary): MutablePythonParameter -> boundary != null }
            .forEach { (name, _, _, _, _, _, boundary) ->
                assert(boundary != null)
                if (boundary!!.isDiscrete) {
                    formattedBoundaries.add(
                        """
                            if not (isinstance($name, int) or (isinstance($name, float) and $name.is_integer())):

                            """.trimIndent()
                            + indent(
                            "raise ValueError('"
                                + name
                                + " needs to be an integer, but {} was assigned."
                                + "'.format("
                                + name
                                + "))"
                        )
                    )
                }
                if (boundary.lowerLimitType !== ComparisonOperator.UNRESTRICTED && boundary.upperLimitType !== ComparisonOperator.UNRESTRICTED) {
                    formattedBoundaries.add(
                        """if not ${boundary.lowerIntervalLimit} ${boundary.lowerLimitType.operator} $name ${boundary.upperLimitType.operator} ${boundary.upperIntervalLimit}:
    """
                            + indent(
                            "raise ValueError('Valid values of "
                                + name
                                + " must be in "
                                + boundary.asInterval()
                                + ", but {} was assigned."
                                + "'.format("
                                + name
                                + "))"
                        )
                    )
                } else if (boundary.lowerLimitType === ComparisonOperator.LESS_THAN) {
                    formattedBoundaries.add(
                        """if not ${boundary.lowerIntervalLimit} < $name:
    """
                            + indent(
                            "raise ValueError('Valid values of "
                                + name
                                + " must be greater than "
                                + boundary.lowerIntervalLimit
                                + ", but {} was assigned."
                                + "'.format("
                                + name
                                + "))"
                        )
                    )
                } else if (boundary.lowerLimitType === ComparisonOperator.LESS_THAN_OR_EQUALS) {
                    formattedBoundaries.add(
                        """if not ${boundary.lowerIntervalLimit} <= $name:
    """
                            + indent(
                            "raise ValueError('Valid values of "
                                + name
                                + " must be greater than or equal to "
                                + boundary.lowerIntervalLimit
                                + ", but {} was assigned."
                                + "'.format("
                                + name
                                + "))"
                        )
                    )
                }
                if (boundary.upperLimitType === ComparisonOperator.LESS_THAN) {
                    formattedBoundaries.add(
                        """if not $name < ${boundary.upperIntervalLimit}:
    """
                            + indent(
                            "raise ValueError('Valid values of "
                                + name
                                + " must be less than "
                                + boundary.upperIntervalLimit
                                + ", but {} was assigned."
                                + "'.format("
                                + name
                                + "))"
                        )
                    )
                } else if (boundary.upperLimitType === ComparisonOperator.LESS_THAN) {
                    formattedBoundaries.add(
                        """if not $name <= ${boundary.upperIntervalLimit}:
    """
                            + indent(
                            "raise ValueError('Valid values of "
                                + name
                                + " must be less than or equal to "
                                + boundary.upperIntervalLimit
                                + ", but {} was assigned."
                                + "'.format("
                                + name
                                + "))"
                        )
                    )
                }
            }
        return formattedBoundaries
    }

    private fun buildParameterCall(): String {
        val formattedParameters: MutableList<String?> = ArrayList()
        val originalNameToValueMap: MutableMap<String, String?> = HashMap()
        pythonFunction.parameters.forEach { (name, defaultValue, assignedBy, _, _, _, _, _, originalDeclaration): MutablePythonParameter ->
            val value: String? =
                if (assignedBy == PythonParameterAssignment.CONSTANT || assignedBy == PythonParameterAssignment.ATTRIBUTE) {
                    defaultValue
                } else {
                    name
                }
            originalNameToValueMap[Objects.requireNonNull(originalDeclaration)!!.name] = value
        }
        Objects.requireNonNull(pythonFunction.originalDeclaration)!!.parameters.stream()
            .filter { (_, _, _, assignedBy): SerializablePythonParameter -> assignedBy != PythonParameterAssignment.IMPLICIT }
            .forEach { (name, _, _, assignedBy): SerializablePythonParameter ->
                if (assignedBy === PythonParameterAssignment.NAME_ONLY) {
                    formattedParameters.add(name + "=" + originalNameToValueMap[name])
                } else {
                    formattedParameters.add(originalNameToValueMap[name])
                }
            }
        return java.lang.String.join(", ", formattedParameters)
    }
}
